packages/automerge-repo/src/Presence.ts

```diff
@@ -0,0 +1,592 @@
+   * peers.
+   *
+   * @param channel
+   * @param msg
+   */
+  broadcast(channel: Channel, msg: State[Channel]) {
+    this.#localState = {
+      ...this.#localState,
+      [channel]: msg,
+    }
+    this.broadcastLocalState()
+  }
+
+  /**
+   * True if {@link dispose} has been called on this Presence and it is no
+   * longer active. Note that a disposed Presence may be started again with
+   * {@link initialize}.
+   */
+  get disposed() {
+    return this.#disposed
+  }
+
+  /**
+   * Dispose of this Presence: broadcast a "goodbye" message (when received,
+   * other peers will immediately forget the sender), stop sending heartbeats,
+   * and stop listening to ephemeral-messages broadcast from peers.
+   *
+   * This can be used with browser events like
+   * {@link https://developer.mozilla.org/en-US/docs/Web/API/Window/pagehide_event | "pagehide"}
+   * or
+   * {@link https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilitychange_event | "visibilitychange"}
+   * to stop sending and receiving updates when not active.
+   */
+  dispose() {
+    if (this.disposed) {
+      return
+    }
+    this.#hellos.forEach(timeoutId => {
+      clearTimeout(timeoutId)
+    })
+    this.#hellos = []
+    this.#handle.off("ephemeral-message", this.#handleEphemeralMessage)
+    this.stopHeartbeats()
+    this.doBroadcast("goodbye")
+    this.#disposed = true
+  }
+
+  private announce() {
+    // Broadcast our current state whenever we see new peers
+    // TODO: We currently need to wait for the peer to be ready, but waiting
+    // some arbitrary amount of time is brittle
+    const helloId = setTimeout(() => {
+      this.broadcastLocalState()
+      this.#hellos = this.#hellos.filter(id => id !== helloId)
+    }, 500)
+    this.#hellos.push(helloId)
+  }
+
+  private broadcastLocalState() {
+    this.doBroadcast("state", { value: this.#localState })
+    // Reset heartbeats every time we broadcast a message to avoid sending
+    // unnecessary heartbeats when there is plenty of actual update activity
+    // happening.
+    this.stopHeartbeats()
+    this.startHeartbeats()
+  }
+
+  private sendHeartbeat() {
+    this.doBroadcast("heartbeat")
+  }
+
+  private doBroadcast(
+    type: PresenceMessageType,
+    extra?: Record<string, unknown>
+  ) {
+    this.#handle.broadcast({
+      userId: this.userId,
+      deviceId: this.deviceId,
+      type,
+      ...extra,
+    })
+  }
+
+  private startHeartbeats() {
+    const heartbeatMs = this.#opts.heartbeatMs ?? DEFAULT_HEARTBEAT_INTERVAL_MS
+    this.#heartbeatInterval = setInterval(() => {
+      this.sendHeartbeat()
+      this.#peers.prune()
+    }, heartbeatMs)
+  }
+
+  private stopHeartbeats() {
+    clearInterval(this.#heartbeatInterval)
+  }
+}
+
+/**
+ * A summary of the latest Presence information for the set of peers who have
+ * reported a Presence status to us.
+ */
+export class PeerPresenceView<State> {
```

Comment:
I'm exposing the current peer Presence state through this class, since (especially with users and devices) some things would be awkward to do with just a raw data structure.